### ISSUE TITLE
Failed to find temporary OS disk on VM.  Please delete manually.

### ISSUE THREAD

**Issue by xpillons** [hashicorp/packer-plugin-azure#190]
#### Overview of the Issue

Packer is failing when building a Windows VM in an existing RG with the following error message 
`Failed to find temporary OS disk on VM.  Please delete manually.`
This makes our pipeline to failed.

#### Reproduction Steps

Try to build an image with the below config file.

### Plugin and Packer version

1.8.0

### Simplified Packer Buildfile

```
{
    "builders": [
        {
            "type": "azure-arm",
            "use_azure_cli_auth": "{{user `var_use_azure_cli_auth`}}",
            "image_publisher": "MicrosoftWindowsDesktop",
            "image_offer": "Windows-10",
            "image_sku": "21h1-pron",
            "image_version": "latest",
            "managed_image_resource_group_name": "{{user `var_resource_group`}}",
            "managed_image_name": "{{user `var_image`}}",
            "os_type": "Windows",
            "vm_size": "Standard_d8s_v3",
            "managed_image_storage_account_type": "Premium_LRS",

            "communicator": "winrm",
            "winrm_use_ssl": true,
            "winrm_insecure": true,
            "winrm_timeout": "5m",
            "winrm_username": "packer",

            "build_resource_group_name": "{{user `var_resource_group`}}",
            "cloud_environment_name": "{{user `var_cloud_env`}}"
        }
    ],
    "provisioners": [{
        "type": "powershell",
        "scripts": [
          "./scripts/windows/wait-for-agent.ps1",
          "./scripts/windows/install-choco.ps1",
          "./scripts/windows/optimize.ps1",
          "./scripts/windows/sysprep.ps1"
        ]
      }]
}

```

### Operating system and Environment details

Running packer from WSL2 - Ubuntu 20.04 

### Log Fragments and crash.log files

```
2022-04-01T18:07:30+02:00: ==> azure-arm: Waiting for deletion of all resources...
2022-04-01T18:07:34+02:00: ==> azure-arm: Failed to find temporary OS disk on VM.  Please delete manually.
==> azure-arm: 
==> azure-arm: VM Name: pkrvmx6hsnxrcvj
==> azure-arm: Error: %!s(<nil>)
2022-04-01T18:07:34+02:00: ==> azure-arm: Removing the created Deployment object: 'kvpkrdpx6hsnxrcvj'
2022-04-01T18:07:51+02:00: ==> azure-arm: 
==> azure-arm: The resource group was not created by Packer, not deleting ...
2022-04-01T18:07:51+02:00: Build 'azure-arm' errored after 9 minutes 11 seconds: Script exited with non-zero exit status: 1.Allowed exit codes are: [0]

```

Set the env var `PACKER_LOG=1` for maximum log detail.

**Comment by invidian** [comment]
I hit the same issue on image-builder CI: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_image-builder/863/pull-azure-sigs/1519012935702155264/artifacts/azure-sigs/windows-2019.log

Plugin actually panicked:
```
[1;31m==> sig-windows-2019: Failed to find temporary OS disk on VM.  Please delete manually.
==> sig-windows-2019: 
==> sig-windows-2019: VM Name: pkrvmh1g2suhr5c
==> sig-windows-2019: Error: %!s(<nil>)[0m
[1;32m==> sig-windows-2019: Removing the created Deployment object: 'kvpkrdph1g2suhr5c'[0m
[1;32m==> sig-windows-2019: 
==> sig-windows-2019: Cleanup requested, deleting resource group ...[0m
[1;32m==> sig-windows-2019: Resource group has been deleted.[0m
[1;31mBuild 'sig-windows-2019' errored after 26 minutes 47 seconds: unexpected EOF[0m

==> Wait completed after 26 minutes 47 seconds

==> Some builds didn't complete successfully and had errors:
--> sig-windows-2019: unexpected EOF

==> Builds finished but no artifacts were created.
panic: runtime error: invalid memory address or nil pointer dereference
2022/04/26 18:33:44 packer-builder-azure-arm plugin: [signal SIGSEGV: segmentation violation code=0x1 addr=0x50 pc=0x1a933ce]
2022/04/26 18:33:44 packer-builder-azure-arm plugin: 
2022/04/26 18:33:44 packer-builder-azure-arm plugin: goroutine 613 [running]:
2022/04/26 18:33:44 packer-builder-azure-arm plugin: github.com/Azure/go-autorest/autorest/azure.(*Future).WaitForCompletionRef(0xc000da6cc0, 0x546d9c0, 0xc000622980, 0x53f6420, 0xc000ab4b00, 0x540e140, 0xc00112f170, 0xc000076620, 0xc000d93080, 0xdf8475800, ...)
2022/04/26 18:33:44 packer-builder-azure-arm plugin: 	/home/circleci/project/packer/vendor/github.com/Azure/go-autorest/autorest/azure/async.go:174 +0x54e
2022/04/26 18:33:44 packer-builder-azure-arm plugin: github.com/hashicorp/packer/builder/azure/arm.(*StepCaptureImage).captureImageFromVM(0xc0007066c0, 0x546d9c0, 0xc000622980, 0xc00071a2a0, 0x18, 0xc000454de0, 0x1c, 0xc0005b9480, 0xc0010302d0, 0x28)
2022/04/26 18:33:44 packer-builder-azure-arm plugin: 	/home/circleci/project/packer/builder/azure/arm/step_capture_image.go:57 +0x225
2022/04/26 18:33:44 packer-builder-azure-arm plugin: github.com/hashicorp/packer/builder/azure/arm.(*StepCaptureImage).Run(0xc0007066c0, 0x546d9c0, 0xc000622980, 0x546e360, 0xc0003e7a10, 0x0)
2022/04/26 18:33:44 packer-builder-azure-arm plugin: 	/home/circleci/project/packer/builder/azure/arm/step_capture_image.go:93 +0x922
2022/04/26 18:33:44 packer-builder-azure-arm plugin: github.com/hashicorp/packer-plugin-sdk/multistep.(*BasicRunner).Run(0xc0003e71d0, 0x546d9c0, 0xc000622980, 0x546e360, 0xc0003e7a10)
2022/04/26 18:33:44 packer-builder-azure-arm plugin: 	/home/circleci/project/packer/vendor/github.com/hashicorp/packer-plugin-sdk/multistep/basic_runner.go:70 +0x234
2022/04/26 18:33:44 packer-builder-azure-arm plugin: github.com/hashicorp/packer/builder/azure/arm.(*Builder).Run(0xc00003ea80, 0x546d9c0, 0xc000622980, 0x54852a8, 0xc0003e7c80, 0x53f8da0, 0xc000410a60, 0x0, 0x0, 0x0, ...)
2022/04/26 18:33:44 packer-builder-azure-arm plugin: 	/home/circleci/project/packer/builder/azure/arm/builder.go:283 +0x14f0
2022/04/26 18:33:44 packer-builder-azure-arm plugin: github.com/hashicorp/packer-plugin-sdk/rpc.(*BuilderServer).Run(0xc00029d800, 0x1, 0xc00038a0e0, 0x0, 0x0)
2022/04/26 18:33:44 packer-builder-azure-arm plugin: 	/home/circleci/project/packer/vendor/github.com/hashicorp/packer-plugin-sdk/rpc/builder.go:117 +0x1c4
2022/04/26 18:33:44 packer-builder-azure-arm plugin: reflect.Value.call(0xc0000c4420, 0xc00000e0b0, 0x13, 0x4dc3f51, 0x4, 0xc000513f08, 0x3, 0x3, 0xc000094000, 0xc000a316f8, ...)
2022/04/26 18:33:44 packer-builder-azure-arm plugin: 	/usr/local/go/src/reflect/value.go:476 +0x8e7
2022/04/26 18:33:44 packer-builder-azure-arm plugin: reflect.Value.Call(0xc0000c4420, 0xc00000e0b0, 0x13, 0xc000a31708, 0x3, 0x3, 0xc000a5f1b8, 0xc000d0e050, 0x462280)
2022/04/26 18:33:44 packer-builder-azure-arm plugin: 	/usr/local/go/src/reflect/value.go:337 +0xb9
2022/04/26 18:33:44 packer-builder-azure-arm plugin: net/rpc.(*service).call(0xc00029dc40, 0xc0000ce0a0, 0xc000462790, 0xc000462a60, 0xc0006cc180, 0xc00013e680, 0x3f2af60, 0xc00038a05c, 0x18a, 0x3e2a5e0, ...)
2022/04/26 18:33:44 packer-builder-azure-arm plugin: 	/usr/local/go/src/net/rpc/server.go:377 +0x189
2022/04/26 18:33:44 packer-builder-azure-arm plugin: created by net/rpc.(*Server).ServeCodec
2022/04/26 18:33:44 packer-builder-azure-arm plugin: 	/usr/local/go/src/net/rpc/server.go:474 +0x44d
```

**Comment by dduportal** [comment]
The Jenkins Project is also having the same error message quite often.

For instance: https://github.com/jenkins-infra/packer-images/pull/288 .

Build log is there: https://gist.github.com/dduportal/fe5954cbe3330d29e805c98bedae8f8d

Packer manifests:
- Windows (crash 2/3 of the builds): https://github.com/jenkins-infra/packer-images/blob/3463fe9d9aede813645527bd4124b7d121cf618f/build-jenkins-agent-windows.pkr.hcl#L12-L27
- Linux (crash 1/4 of the builds): https://github.com/jenkins-infra/packer-images/blob/3463fe9d9aede813645527bd4124b7d121cf618f/build-jenkins-agent-ubuntu.pkr.hcl#L11-L20

**Comment by JenGoldstrich** [comment]
Hey @xpillons thanks for submitting this issue!

Judging @invidian's logs I can see that the panic seems to be occurring from an upstream Azure SDK library called [`go-autorest`](https://github.com/Azure/go-autorest), specifically at 

```
2022/04/26 18:33:44 packer-builder-azure-arm plugin: github.com/Azure/go-autorest/autorest/azure.(*Future).WaitForCompletionRef(0xc000da6cc0, 0x546d9c0, 0xc000622980, 0x53f6420, 0xc000ab4b00, 0x540e140, 0xc00112f170, 0xc000076620, 0xc000d93080, 0xdf8475800, ...)
2022/04/26 18:33:44 packer-builder-azure-arm plugin: 	/home/circleci/project/packer/vendor/github.com/Azure/go-autorest/autorest/azure/async.go:174 +0x54e
2022/04/26 18:33:44 packer-builder-azure-arm plugin: github.com/hashicorp/packer/builder/azure/arm.(*StepCaptureImage).captureImageFromVM(0xc0007066c0, 0x546d9c0, 0xc000622980, 0xc00071a2a0, 0x18, 0xc000454de0, 0x1c, 0xc0005b9480, 0xc0010302d0, 0x28)
```

The Packer team will need to reach out to some folks on the Microsoft side because, Azure go-autorest is set to be removed from support as of [March 31st 2023](https://azure.microsoft.com/updates/support-for-azure-sdk-libraries-that-do-not-conform-to-our-current-azure-sdk-guidelines-will-be-retired-as-of-31-march-2023/).


@invidian @KshitijAgrawal if y'all have any insight into the removal of support for go-autorest and other go Azure Go SDKs please chime in.

**Comment by lemeurherve** [comment]
Hello, any update about this issue?

**Comment by JenGoldstrich** [comment]
Thanks for reaching out for updates. We are working internally and with our contacts at Microsoft to understand the steps forward here. Auto-rest (the Azure dependency that is panicing here https://github.com/hashicorp/packer-plugin-azure/issues/190#issuecomment-1218437340) is being deprecated this year. So we looking into making the necessary updates to move away from it, as patching it is not feasible at this time. We will post any updates about this process in the SDK Upgrade post. https://github.com/hashicorp/packer-plugin-azure/issues/243

**Comment by vanmash** [comment]
hey,
any chance for a workaround or permanent fix?

**Comment by JenGoldstrich** [comment]
Hey all, reviewing this issue again it looks like there are two errors happening here 
The first issue is that the plugin outputs `Failed to find temporary OS disk on VM please delete manually` during builds.  Digging into this I found we rely on the same Cleanup function for VM deploys and KeyVault deploys.  The KeyVault deploy Cleanup tries to delete a VM Disk that doesn’t exist.  This is why we’re seeing this errant output.  I’ve cleaned this output up as part of https://github.com/hashicorp/packer-plugin-azure/pull/296.  This error never caused the plugin to stop executing, it just outputs the error and short circuits the cleanup function.

The second issue is the panic from autorest.   Looking at the [CI output](https://github.com/hashicorp/packer-plugin-azure/issues/190#issuecomment-1193100426) I see this error. ```[2022-07-23T09:23:03.768Z] 2022-07-23T09:23:03Z: ==> azure-arm.ubuntu: ERROR: -> PropertyChangeNotAllowed : Changing property 'sourceVirtualMachine' is not allowed.```.  This is the actual error thats happening when trying to capture the managed image in this build.  The Azure API is returning an error and the SDK/autorest are panicking from this API error.  This specific error happens presumably during parallel builds when trying to modify the same Managed Image.  Which I could see happening in CI environments.  We can not change the API error being thrown as Azure does not allow existing Managed Images to have their source VM changed, we can only handle this error better in the future.   
If you think I’m mistaken can you please post your entire CI log/simplified Packer template and we can dig into specific errors y’all are seeing? I think this panic might occur in several places whenever the SDK incorrectly handles Azure errors.

Potential workarounds for this you could be to try to direct publish to Compute Gallery instead as that will skip publishing a managed image, appending a random string to the end of the managed image name to allow multiple builds, or restricting your CI to one build at a time.  I appreciate y’alls patience on this, and hope we can find a clean resolution!

**Comment by dduportal** [comment]
Hello @JenGoldstrich thanks for this check and feedback!

> The second issue is the panic from autorest.

We were able to apply the change you describe as part of building ARM64 Linux VM in Azure, which forbids creating a managed image.

Combined with https://github.com/jenkins-infra/packer-images/pull/609 where we generate a unique gallery version for each build (to avoid concurent deployment to the gallery which also triggers the error `PropertyChangeNotAllowed : Changing property 'sourceVirtualMachine' is not allowed`), we achieved a clean process with no more this error .


Thanks for #296 that will also help a lot!
